apps/yggdrasil/test/yggdrasil/user_test.exs

```diff
@@ -194,4 +338,147 @@ defmodule Yggdrasil.UserTest do
+
+    assert User.is_granted? user, res_uniq_perm_list
+  end
+
+  test "is_granted?/2 returns true for any one resources's granted permission set" do
+    {user, res_uniq_perm_list, _} = user_perms_for_roles(@test_roles)
+
+    Enum.each res_uniq_perm_list, fn res_perm ->
+      assert User.is_granted? user, [res_perm]
+    end
+  end
+
+  test "is_granted?/2 returns true any combination of granted permissions" do
+    {user, res_uniq_perm_list, _} = user_perms_for_roles(@test_roles)
+
+    {res_perm_combo, uniq_res_perm_combos} = produce_resoure_perms_combos(res_uniq_perm_list)
+
+    # lets check each resource combo individually first
+    Enum.each res_perm_combo, fn res_list ->
+      Enum.each res_list, fn combo ->
+        assert User.is_granted? user, [combo]
+      end
+    end
+
+    # now lets check all the combos
+    Enum.each uniq_res_perm_combos, fn combo ->
+      assert User.is_granted? user, combo
+    end
+  end
+
+  test "is_granted?/2 returns false for all combinations of non granted permissions" do
+    {user, res_uniq_perm_list, _} = user_perms_for_roles(@test_roles)
+
+    res_uniq_perm_list = Enum.map res_uniq_perm_list, fn {res, perms} ->
+      perms = Enum.map perms, fn p -> 
+        p = Atom.to_string(p)
+        p = p <> "_false"
+        String.to_atom(p)
+      end
+
+      {res, perms}
+    end
+
+    {res_perm_combo, uniq_res_perm_combos} = produce_resoure_perms_combos(res_uniq_perm_list)
+
+    # lets check each resource combo individually first
+    Enum.each res_perm_combo, fn res_list ->
+      Enum.each res_list, fn combo ->
+        refute User.is_granted? user, [combo]
+      end
+    end
+
+    # now lets check all the combos
+    Enum.each uniq_res_perm_combos, fn combo ->
+      refute User.is_granted? user, combo
+    end
+  end
+
+  test "create_with_default_role/1 creates user and assigns 'player' as the default role" do
+    role = Repo.one!(from r in Role, where: r.name == "player", select: r)
+
+    {:ok, user} = User.create_with_default_role(@valid_attrs) 
+    user = Repo.preload(user, :user_roles)
+
+    default_role = Enum.find user.user_roles, fn ur ->
+      ur.role_id == role.id
+    end
+
+    assert default_role.role_id == role.id
+  end
+
+  test "assign_role/2 associates supplied role to user" do
+    [role|_rest] = insert_test_roles
+
+    user = %User{}
+    |> User.create_changeset(@valid_attrs)
+    |> Repo.insert!
+
+    :ok = User.assign_role(user, role.name)
+
+    user = Repo.preload(user, :user_roles)
+
+    assigned_role = Enum.find user.user_roles, fn ur ->
+      ur.role_id == role.id
+    end
+    assert assigned_role.role_id == role.id
+  end
+
+  test "assign_role/2 with invalid user returns an error" do
+    [role|_rest] = insert_test_roles
+
+    user = %User{id: -12}
+
+    assert {:error, msg} = User.assign_role(user, role.name)
+    assert msg.errors == [user: "does not exist"]
+  end
+
+  test "assign_role/2 with invalid role raises an Ecto.NoResultsError" do
```

Comment:
I think we want this to crash here, there should never be supplied role by name from the client, and we would defaulted this wrong in code if it blows up or the role doesn't exist one.

In hindsight I don't think that should not look up the role by name though. Some of this will shake out the more we use it. I think there is more elegant way to do this, just haven't had time to really explore that.